Improve test suite

gonzaponte opened the pull request:
Following #1, I implemented a bunch of tests that were missing. I have also taken the chance to extend some of the preexisting tests and to slightly document some features.

Closes #1.

gonzaponte committed:
Add test for implicit pipes in fork

gonzaponte committed:
Add tests for count_filter

gonzaponte committed:
Add test for spy_count

gonzaponte committed:
Expand push futures tests to consider all features

gonzaponte committed:
Add tests for string_to_pick

gonzaponte committed:
Implement test_pipes_must_end_in_a_sink

gonzaponte committed:
Add test to ensure branch closes sideways

gonzaponte committed:
Introduce bug to demonstrate test catches it

gonzaponte committed:
Revert "Introduce bug to demonstrate test catches it"

This reverts commit f80567703536fab318da98d1133934cacf545bd8.

gonzaponte commented:
I am particularly unsure about `test_string_to_pick_ignores_components`. I am not sure it is even worth to test that. In case it is, I am not sure how to do it without getting into the mess of writing a test for each case.

jacg commented on dataflow_test.py:
```diff
@@ -111,6 +112,27 @@ def test_longer_pipeline():
+
+    df.push(source = the_source,
+            pipe   = df.fork(pipe_as_tuple,
+                             pipe_as_pipe ))
```
Nice test.

I'm just wondering whether `implicit_pipe` and `explicit_pipe` might be better names than `pipe_as_tuple` and `pipe_as_pipe`. Or maybe `implicit_pipe_as_tuple`.

jacg commented on dataflow_test.py:
```diff
@@ -452,3 +452,33 @@ def test_slice_raises_ValueError(args):
+def test_count_filter_filters_correctly():
+    even_numbers = list(range(0, 20, 2))
+    odd_numbers  = list(range(1, 20, 2))
+    the_source   = odd_numbers + even_numbers
```
It makes me a little nervous that they are not interleaved at all.

gonzaponte replied:
I thought about that... given the simplicity of the filtering maybe it is enough to use the continuous sequence and split them for the comparison afterwards

jacg commented on dataflow_test.py:
```diff
@@ -452,3 +452,33 @@ def test_slice_raises_ValueError(args):
+    assert filtered == odd_numbers
+
+
+def test_count_filter_counts_correctly():
```
I'm tempted to unite `test_count_filter_{counts,filters}_correctly`.

+ The two aspects are very closely tied in this component
+ There's lots of repetition between the two tests, making for a suboptimal signal to noise ratio.

jacg commented on dataflow_test.py:
```diff
@@ -280,6 +280,21 @@ def test_spy():
+def test_spy_count():
+    the_source = list(range(20))
+
+    count     = df.count()
```
Nice test, but the customary explanatory comment would be useful.

jacg commented on dataflow_test.py:
```diff
@@ -168,17 +168,26 @@ def test_count():
+# 'push' provides a higher-level interface to using such futures:
+# it optionally accepts a future, a tuple of futures or a mapping
+# of futures. It returns the result of the future, a tuple of
+# their results or a namespace with the results, respectively.
```
I'll need to take some time to calmly think about whether we can unite these three tests. The crucial differences seem drowned in the noise of the setup ... but I don't immediately see how this can be mitigated.

jacg commented on dataflow_test.py:
```diff
@@ -523,3 +523,71 @@ def test_count_filter_counts_correctly():
+              df.branch(df.sink(print)) ,
+              df.pipe  (df.map(abs))    ))
+def test_string_to_pick_ignores_components(component):
+    assert component is df.string_to_pick(component)
```
Two points:

1. I don't think that `string_to_pick` is part of the public API (so it should start with an underscore!). If that is the case, then it shouldn't be tested here, because these are documentation tests aimed at users.

2. As you well know, I love `parametrize`, but I'm reluctant to use it in user-facing documentation tests, because a complete layman should be able to read these and understand pretty much anything. `parametrize` will freak them out.

jacg commented on dataflow_test.py:
```diff
@@ -523,3 +523,71 @@ def test_count_filter_counts_correctly():
+def test_string_to_pick():
+
+    # string_to_pick creates a pipe component that picks
+    # an item from the namespace and pushes it through the pipe
```
This comment assumes that the user knows that there is (well ... might be) a namespace flowing through the pipe. I don't think this has been mentioned anywhere yet.

In general, I think that we'll need to think carefully about how we introduce the idea of the namespace flowing though the pipes, and all the features that exist purely for working with these namespaces.

gonzaponte replied:
Well, `string_to_pick` is specifically introduced to work with that feature, so it must assume it exists, right? But the documentation needs to be there beforehand, I agree.

jacg replied:
Yup. I think there are a few things to be done here:

1. Don't mention `string_to_pick` at all in the user-facing docs (which these tests are, at least for the time being), only show the use of strings at the front of pipes.

2. Ensure that the namespace flow is discussed before this.

3. Rename `string_to_pick` to `_string_to_pick` (underscore).

4. Test `_string_to_pick` in a non-user-facing unit tests, if deemed necessary.

5. Think about whether we want to admit tuples-to-pick in addition to a single string.

gonzaponte replied:
OK. Do you want to address all that in this PR? I would certainly leave the last point for a different PR.

jacg replied:
Given that I'm still going to be rebasing `master` like crazy, for a while, I'm reluctant to accept any PR for the time being. So we have some time to think about this.

The last point is definitely not for this PR, agreed. The others, might make sense in here.

jacg commented on dataflow_test.py:
```diff
@@ -523,3 +523,71 @@ def test_count_filter_counts_correctly():
+    # an item from the namespace and pushes it through the pipe
+
+    the_source_elements = list(range(10))
+    the_source          = (dict(x=i) for i in the_source_elements)
```
I'd be tempted to populate it with more attributes than just the one that you are going to extract.

jacg commented on dataflow_test.py:
```diff
@@ -523,3 +523,71 @@ def test_count_filter_counts_correctly():
+
+# string_to_pick is not meant to be called directly.
+# It's action is implicit when the first element of
+# a pipe is a string. This also works with forks
```
Excactly: this agrees with my earlier comment about `string_to_pick` not being part of the API, and therefore having to place in these user-facing documentation tests.

jacg commented on dataflow_test.py:
```diff
@@ -490,9 +490,13 @@ def test_slice_raises_ValueError(args):
+    the_source    = range(10)
+    sinkless_pipe = df.map(abs)
+
+    with raises(AttributeError):
```
We should make some effort to make the message in this error be as helpful as possible. Or even raise our own specific exception type.

jacg replied:
Perhaps this should be listed in an issue which also mentions the points that @mmkekic made (in the thread in IC) about error clarity.

jacg commented on dataflow_test.py:
```diff
@@ -357,6 +357,19 @@ def test_branch():
+                             the_main_sink))
+
+    with raises(StopIteration):
+        the_branch_sink.send(99)
```
Hmm ... is this enough? The test name implies (at least to me) that closing the branch should close other things which would otherwise remain open. I don't see the body of the test addressing that kind of stuff at all.

I'll need to think about this more carefully. Don't have time now.

[Aside: before I forget ... I suspect that the `close_all` option in `slice` is not tested adequately]

gonzaponte replied:
Well, this test was particularly hard to develop because I am not yet familiar with the whole coroutine thing. The idea was that originally we discovered that when doing a branch the sideways pipeline was not closed at the end of a push, but I don't remember how or what implication it had. So I started playing with it a bit and I realized that not closing that pipeline means that you could still send data through it, so that's what I attempted here.

I think in order to write a good test for this I will need your help, so we can wait until you find the time for it.

> Aside: before I forget ... I suspect that the close_all option in slice is not tested adequately

Yes. I've found a comment in the code saying so, I will add it to the list.

jacg replied:
Whilst you're waiting for me, and only if you feel like it, try

1. Checking the *other* branch. At the moment you're checking the side-branch, but not the main one.

Did your reverted test-failure-provocation make anything fail?

One likely consequence of some branch not being closed would be any future associated with the sink at the end of the branch, not having a value available.

gonzaponte committed:
Make `string_to_pick` private

gonzaponte committed:
Add tests for working with a namespace

gonzaponte committed:
Raise exception if the pipe is not complete

gonzaponte committed:
Add test for slice with `close_all=True|False`

gonzaponte committed:
Fix bug in slice with `close_all=True`

In the last iteration there was nothing preventing from raising
StopPipeline before the iteration was completed. An empty `yield`
expression is added to force one more iteration.

gonzaponte commented:
For now, I have made the changes adding new commits. When we are happy I can rewrite the history.

jacg commented on dataflow_test.py:
```diff
@@ -114,23 +114,23 @@ def test_longer_pipeline():
-    # Arguments can be pipes or tuples
-    # tuples get implicitely converted into pipes
+    # Arguments can be pipes or tuples.
+    # Tuples get implicitely converted into pipes
```
'implicitly' without the E

jacg commented on dataflow_test.py:
```diff
@@ -114,23 +114,23 @@ def test_longer_pipeline():
-            pipe   = df.fork(pipe_as_tuple,
-                             pipe_as_pipe ))
+            pipe   = df.fork(implicit_pipe,
+                             explicit_pipe))
```
I think that implicit pipes are designed to reduce noise and number of characters typed in more complex graphs. So I don't think that you would ever write something like line 126 in real life, and I don't think we should encourage it over line 127. So, while I like the clarity of this test as you've written it, as a user-facing documentation test, I think I would prefer to combine 126-131 into

````python
df.push(source = the_source,
        pipe   = df.fork(       (add_1, implicit_pipe_sink),
                         df.pipe(add_1, explicit_pipe_sink))
````

Looking at it, I think that the sink names in there, get the point across quite well. What do you think?

Also, off the top of my head, I don't remember whether `push(pipe = ...)` accepts implicit pipes, and whether it's tested or not. If it doesn't it probably should.

gonzaponte replied:
>  think I would prefer to combine 126-131 into
> ``` df.push(source = the_source,
>        pipe   = df.fork(       (add_1, implicit_pipe_sink),
>                         df.pipe(add_1, explicit_pipe_sink))
> ```
> Looking at it, I think that the sink names in there, get the point
> across quite well. What do you think?

I completely agree. It looks much better like that.

> Also, off the top of my head, I don't remember whether `push(pipe = ...)` accepts implicit pipes, and whether it's tested or not. If it doesn't it probably should.

I don't remember either. Will check and test if necessary.

gonzaponte replied:
> I don't remember whether push(pipe = ...) accepts implicit pipes

It doesn't. Do you want it to? If so, do you want to include it in this PR?

jacg replied:
> Do you want it to?

I think that the `pipe` keyword argument makes it abundantly clear that it's a pipe, so the `df.pipe` that immediately follows it is just boilerplate/noise and needlessly increases the indentation of the graph by 7 or 4 characters. So I'd say yes.

I trust that the pipe argument is keyword-only ... hmm, it isn't. Do you agree that it (well, all three of `pipe`s admissible arguments) should be keyword-only? Hmm, maybe not. Need to ruminate on this.

> If so, do you want to include it in this PR?

No. Let's try to keep them small and orthogonal. This is already turning into a monster.

gonzaponte replied:
> It is just boilerplate/noise and needlessly increases the indentation of the graph by 7 or 4 characters

I agree.

> Do you agree that it (well, all three of `push`'s admissible arguments) should be keyword-only?

I don't see any reason why not. It encourages (actually forces) a verbose and obvious-to-the-eye syntax. I am quite comfortable with making it keyword-only.

> No. Let's try to keep them small and orthogonal.

Good. @mmkekic told me she wanted to contribute. Maybe this is a good introduction point.

jacg replied:
OK, executive decision: `pipe`'s arguments should be keyword-only. Not in this PR.

>  @mmkekic told me she wanted to contribute.

Great!

> Maybe this is a good introduction point.

You mean get her to make these args keyword-only? It's literally adding 3 characters (including a space), and I'm confident that she's familiar with the style in which we'll be running this project, so it will be trivial for her and not especially enlightening ... but I have no objections. My only reservation is that I don't want to accept PRs until I'm happy that I won't be rebasing master any more ...

gonzaponte replied:
> You mean get her to make these args keyword-only? 

No, I meant adding the implicit-pipe feature to the pipe argument. It's not a big deal either, but it will require a test and it's a bit more elaborate.

jacg replied:
Yup, excellent.

jacg commented on dataflow_test.py:
```diff
@@ -513,17 +513,16 @@ def test_pipes_must_end_in_a_sink():
-    odd_numbers  = list(range(1, 20, 2))
-    the_source   = odd_numbers + even_numbers
+    the_source  = list(range(21))
+    predicate   = lambda n: n % 2
```
I like the appearance of the name 'predicate'.

jacg commented on dataflow_test.py:
```diff
@@ -513,17 +513,16 @@ def test_pipes_must_end_in_a_sink():
             pipe   = df.pipe(odd.filter, the_sink))
 
-    assert filtered == odd_numbers
+    assert filtered == list(filter(predicate, the_source))
```
Yup, I don't think we should make any secret or apology about the similarity between our `map`, `filter` and `reduce` and the standard ones. So I'd say it absolutely right to highlight these similarities in the tests.

jacg commented on dataflow_test.py:
```diff
@@ -512,33 +512,21 @@ def test_pipes_must_end_in_a_sink():
+    expected_result = list(filter(predicate, the_source))
+    assert filtered        == expected_result
+    assert result.n_passed == len(expected_result)
+    assert result.n_failed == len(the_source) - len(expected_result)
```
Align `expected_result` on these 3 lines?

I think these two tests are much better united!

jacg commented on dataflow_test.py:
```diff
@@ -513,6 +513,11 @@ def test_pipes_must_end_in_a_sink():
+
+    # count_filter provides a future/filter pair.
+    # This is a simple interface to keep track of
+    # how many entries satisfy the predicate.
```
Maybe mention that it not only keeps track of how many pass, but also how many were filtered out (which I had forgotten myself).

jacg commented on dataflow_test.py:
```diff
@@ -307,6 +307,13 @@ def test_spy():
+    # of a pipeline. However, because it is a sink it needs
+    # to be plugged into a spy. Thus, the component spy_count
+    # provides a comfortable interface to access the future
+    # and spy objects in a single line.
```
I think that we should replace all of these by a single, pithily-named, sink-to-spy converter.

As for the wording ... well, let's talk about it when we get there.

gonzaponte replied:
I guess you mean futuresink-to-futurespy? sink-to-spy would just be `df.branch`.

jacg replied:
Yup.

Makes keeping the name pithy a bit more difficult, though.

Note to self: sink-to-spy = df.branch needs to be an example in the documentation somewhere, if it isn't already.

gonzaponte replied:
Suggestions for the name? I think `sink_to_spy` is misleading. `fsink_to_fspy` is not as obvious as I would like.

jacg replied:
Let me start writing the wordy documentation (which I've been on the verge of doing for the last 2 hours), and see what inspiration I get.

Have you tried to get mdbook working locally? The Rust package situation is very good indeed, so you should have no problem. It will compile it from source, so it just takes a while.

It would be good if you could contribute to that too. I suspect that making some progress over there will clear up all sorts of things in terms of what we want to do and say in this PR.

gonzaponte replied:
I haven't got into that yet. Will take a look.

jacg commented on dataflow_exhaustive_test.py:
```diff
@@ -0,0 +1,29 @@
+from pytest import mark
```
How about calling the original test file `documentation_test.py` or `documenting_test.py`, and this this one ... well, *dataflow* is superfluous in the name, now that dataflow is all that there is. *Exhaustive* works. Maybe *robust*.

gonzaponte replied:
Yes, I was thinking about renaming this test file, but I forgot to mention it yesterday. What do you prefer, a first commit changing the name of the file or a last commit renaming them?

jacg replied:
No strong opinion. As you know, I like a clean history, which leans me towards getting the name right in the first place. OTOH I don't want to mess with the historical commits that came from IC, so we're stuck with the wrong name at the beginning of history. At which point it really doesn't matter when the name changes. Do whatever is easiest: don't waste any time on Git-fiddling on this.

jacg commented on dataflow_test.py:
```diff
@@ -541,6 +543,70 @@ def test_count_filter():
+# arguments that allow the user to specify which node of
+# information the component should use. The output of
+# the component can be put back under the same or a
+# different label.
```
The motivation behind the namespaces-in-the-stream was that it was the simplest way of enabling parallel, synchronized and joinable streams. (It's a bit of a hack, but it's a damn simple, effective, flexible and easy to understand hack. I find it very unsatisfying from the purity perspective: we're *mutating* (ugh!) a namespace; they're not *really* flow networks/graphs ... but it's very pragmatic.)

Anyway, the documentation of this feature should highlight this parallel, synchronized, joinable flows perspective.

This will require careful thought, and a whole bunch of non-trivial examples that emphasize this aspect.

jacg commented on dataflow.py:
```diff
@@ -247,9 +247,13 @@ def stop_when_loop(item):
 
-class StopPipeline(Exception): pass
+class StopPipeline       (Exception): pass
+class IncompletePipeError(Exception): pass
```
I'm not a fan of naming exceptions `*Error`. For some it is more reasonable than others. For example, I think that `KeyError` is completely misnamed as an error: it is not an error to ask a question to which the answer happens not to be known right now. If the answer *cannot* be known, `ZeroDivisionError` for example, then it would be OK to call it an error.

Broadly speaking, exceptions are messages about some noteworthy situations. I'd avoid calling them errors unless it being an error is a fundamental part of the message.

I'd have to think carefully before deciding about this particular message.

jacg commented on dataflow_test.py:
```diff
@@ -498,7 +498,20 @@ def test_slice_downstream(spec):
 
 
-#TODO: Write test slice_close_all
+def test_slice_close_all():
```
As these are user-facing documentation tests, I think that it's important that the difference between `close_all = True/False` should be made clear, without the reader having to hunt around.

jacg commented:
I'm going to need to add [anchors](https://rust-lang.github.io/mdBook/format/mdbook.html#including-portions-of-a-file) to the documentation test source. So it would be good if we could merge your fine work ASAP, before doing that.

Let's just clean up the history of what we have got so far, and merge. Then I think we'll move the whole documenting test source into the doc directory, and the thing you called `exhaustive_test.py` can simply be called `dataflow_test.py` and we can go to town on putting exhaustive tests in there. The `slice` tests should be moved there.

gonzaponte commented:
Ok!

gonzaponte commented:
I've reduced it to 14 commits. What about the points that were not dealt with yet? Do we leave them for another PR?

gonzaponte committed:
Remove TODOs

jacg commented:
Yup. Let's get this in, so that the doc generation can be built on top of it.

jacg commented on dataflow_test.py:
```diff
@@ -495,7 +495,32 @@ def test_slice_downstream(spec):
+        assert result_main   == the_source[:n_elements]
+    else:
+        assert result_branch == the_source[:n_elements]
+        assert result_main   == the_source
```
Very nice!
